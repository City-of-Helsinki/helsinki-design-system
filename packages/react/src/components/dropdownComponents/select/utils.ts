import React from 'react';

import { SelectData, SelectMetaData, SelectDataHandlers, FilterFunction } from './types';
import {
  Group,
  Option,
  ModularOptionListData,
  ModularOptionListProps,
  ScreenReaderNotification,
} from '../modularOptionList/types';
import { ChangeEvent } from '../../dataProvider/DataContext';
import { eventTypes } from './events';
import {
  iterateAndCopyGroup,
  getGroupLabelOption,
  getAllOptions,
  propsToGroups,
  getSelectedOptions,
} from '../modularOptionList/utils';

// Re-export commonly used utility functions
export {
  getAllOptions,
  getSelectedOptions,
  propsToGroups,
  updateGroupLabelAndOptions,
  updateOptionInGroup,
  defaultFilter,
} from '../modularOptionList/utils';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function createScreenReaderNotification(type: string, content: string, delay = 0): ScreenReaderNotification {
  return {
    type,
    content,
    delay,
    showTime: 0,
    addTime: Date.now(),
  };
}

export function filterOptions(groups: ModularOptionListData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function filterSelectableOptions(
  groups: ModularOptionListData['groups'],
  filterStr: string,
  filterFunc: FilterFunction,
  isMultiSelect: boolean,
) {
  return getAllOptions(groups, !isMultiSelect).filter((option) => filterFunc(option, filterStr));
}

export function mergeSearchResultsToCurrent(
  props: Pick<ModularOptionListProps, 'groups' | 'options'>,
  currentGroups: ModularOptionListData['groups'],
): ModularOptionListData['groups'] {
  const newData = propsToGroups(props) || [];
  const newOptions = getAllOptions(newData);
  const currentOptionsWithoutMatches = getSelectedOptions(currentGroups).filter((option) => {
    const sameInNewOptionsIndex = newOptions.findIndex((newOption) => {
      return newOption.value === option.value;
    });
    if (sameInNewOptionsIndex > -1) {
      newOptions[sameInNewOptionsIndex].selected = true;
      return false;
    }
    return true;
  });

  const currentHiddenOptionsInAGroup = currentOptionsWithoutMatches.length
    ? [{ options: currentOptionsWithoutMatches.map((opt) => ({ ...opt, visible: false })) } as Group]
    : [];

  return [...currentHiddenOptionsInAGroup, ...newData];
}

/**
 *
 * @param notification
 * @param dataHandlers
 * @returns Returns true, if notification was added
 */
export function addOrUpdateScreenReaderNotificationByType(
  notification: ScreenReaderNotification,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfSameType = screenReaderNotifications.findIndex((n) => n.type === notification.type);
  if (indexOfSameType > -1) {
    const updatedList = [...screenReaderNotifications];
    const prev = updatedList[indexOfSameType];
    if (prev.content === notification.content) {
      return false;
    }
    updatedList[indexOfSameType] = notification;
    dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
    return false;
  }
  const updatedList = [...screenReaderNotifications, notification];
  dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
  return true;
}

export function removeScreenReaderNotification(
  target: Partial<ScreenReaderNotification>,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfMatch = screenReaderNotifications.findIndex((n) => {
    const hasTypeMatch = !target.type || n.type === target.type;
    const hasContentMatch = !target.content || n.content === target.content;
    return hasTypeMatch && hasContentMatch;
  });
  if (indexOfMatch > -1) {
    screenReaderNotifications.splice(indexOfMatch, 1);
    dataHandlers.updateMetaData({ screenReaderNotifications });
    return true;
  }
  return false;
}
