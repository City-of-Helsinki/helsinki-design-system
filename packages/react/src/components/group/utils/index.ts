import {
  SyntheticEvent,
  FunctionComponent,
  HtmlHTMLAttributes,
  PropsWithChildren,
  ComponentType,
  MutableRefObject,
} from 'react';

import { StorageData, Storage } from './storage';

export type ChangeEventPayload = { value?: unknown; originalEvent?: SyntheticEvent };
export type ChangeEvent = { id: string; type?: string; payload?: ChangeEventPayload };
export type ChangeHandlerProps = ChangeEvent & { controller: Controller };
export type ChangeHandler = (props: ChangeHandlerProps) => unknown;
export type HTMLElementAttributesWithChildren = PropsWithChildren<HtmlHTMLAttributes<HTMLElement>>;
export type DefaultGroupElementProps = Record<string, unknown> & {
  key?: string;
  forwardedController?: Controller;
  'data-hds-group-id'?: string;
};

export type ElementRef = MutableRefObject<HTMLElement | null>;

export type Controller = {
  getProps: ({ id: string, elementProps: DefaultGroupElementProps }) => ReturnType<PropSetter>;
  updateData: (props: { data: StorageData; updateKeys?: string[] }) => void;
  getData: Storage['get'];
  getMetaData: () => StorageData;
  updateMetaData: (appended: StorageData) => StorageData;
  triggerChange: (change: ChangeEvent) => unknown;
};

export type ControllerProps = {
  reRenderer: () => void;
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange: ChangeHandler;
};

export type PropSetter<T = DefaultGroupElementProps> = (props: {
  id: string;
  controller: Controller;
  key: string;
  elementProps: DefaultGroupElementProps;
}) => T;

export type PropHandler = {
  getProps: (props: {
    id: string;
    controller: Controller;
    elementProps: DefaultGroupElementProps;
  }) => DefaultGroupElementProps;
  updateKey: (id: string) => void;
};

export type GroupChildFunction = FunctionComponent<{ controller: Controller }>;

export type GroupChild<P = DefaultGroupElementProps> = ComponentType<HTMLElementAttributesWithChildren & P>;

export type GroupProps = PropsWithChildren<unknown> & {
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange?: ChangeHandler;
};
