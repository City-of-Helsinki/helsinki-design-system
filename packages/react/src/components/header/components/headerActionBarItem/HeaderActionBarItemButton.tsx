import React, { cloneElement, forwardRef, ReactNode, RefObject } from 'react';

import classes from './HeaderActionBarItemButton.module.scss';
import classNames from '../../../../utils/classNames';
import { IconAngleDown, IconAngleUp } from '../../../../icons';

type ButtonAttributes = JSX.IntrinsicElements['button'];

export interface HeaderActionBarItemButtonProps extends ButtonAttributes {
  /**
   * Fix the item position to the right side of the action bar.
   */
  fixedRightPosition?: boolean;
  /**
   * Icon element for the action bar item.
   */
  icon?: ReactNode;
  /**
   * Label for the action bar item.
   */
  label?: string | JSX.Element;
  /**
   * Positions the label to the right side of the icon.
   * @internal
   */
  labelOnRight?: boolean;
  /**
   * Label shown only visually when button is in active state. Screenreaders see only the "label".
   */
  activeStateLabel?: string | JSX.Element;
  /**
   * Icon shown when button is in active state.
   */
  activeStateIcon?: ReactNode;
  /**
   * Indicates button is in active state
   */
  isActive?: boolean;
  /**
   * Possibility to use a full-width version of the dropdown, for example in mobile use.
   */
  fullWidth?: boolean;
  /**
   * If SubItems
   */
  hasSubItems?: boolean;
  /**
   * Initials for avatar which replace icon.
   */
  avatar?: string | JSX.Element;
  ref?: RefObject<HTMLButtonElement>;
  /**
   * Menu button resizing is prevented by rendering button's active state to a separate element.
   */
  preventButtonResize?: boolean;
}

export const HeaderActionBarItemButton = forwardRef<HTMLButtonElement, HeaderActionBarItemButtonProps>(
  (
    {
      icon,
      label,
      labelOnRight,
      fixedRightPosition,
      className,
      activeStateLabel,
      activeStateIcon,
      isActive,
      fullWidth,
      hasSubItems,
      avatar,
      preventButtonResize,
      ...rest
    },
    ref,
  ) => {
    const hasActiveState = !!(activeStateLabel || activeStateIcon);
    const buttonClassName = classNames({
      [classes.actionBarItemButton]: true,
      [className]: true,
      [classes.fixedRightPosition]: fixedRightPosition,
      [classes.isActive]: hasActiveState && isActive,
      [classes.fullWidth]: fullWidth,
      [classes.hasSubItems]: hasSubItems,
      [classes.preventButtonResize]: preventButtonResize && hasActiveState && isActive,
    });
    const iconClassName = classNames({ [classes.actionBarItemButtonIcon]: true, [classes.labelOnRight]: labelOnRight });
    const labelClassName = classNames({
      [classes.actionBarItemButtonLabel]: true,
      [classes.labelOnRight]: labelOnRight,
    });

    const Icon = ({
      element,
      elementClassName,
    }: {
      element: HeaderActionBarItemButtonProps['icon'];
      elementClassName?: string;
    }) => {
      if (!element && !React.isValidElement(element)) {
        return null;
      }
      return (
        <span className={elementClassName}>{cloneElement(element as React.ReactElement, { 'aria-hidden': true })}</span>
      );
    };

    const Label = ({
      text,
      isForActiveState,
    }: {
      text: HeaderActionBarItemButtonProps['label'];
      isForActiveState?: boolean;
    }) => {
      return (
        <span
          className={classNames(labelClassName, isForActiveState && classes.activeStateContentLabel)}
          {...(isForActiveState && !isActive ? { 'aria-hidden': true } : {})}
        >
          {text}
        </span>
      );
    };

    const showActiveState = hasActiveState && isActive && (!fullWidth || !hasSubItems) && !preventButtonResize;
    const showOpenButton = hasSubItems && !isActive;
    const showCollapseButton = hasSubItems && isActive;
    const showAvatar = avatar !== undefined && !showActiveState;
    const showLabel = (preventButtonResize && hasActiveState) || (!(showAvatar && !fullWidth) && !showActiveState);
    const showIcon =
      icon !== undefined && avatar === undefined && ((preventButtonResize && hasActiveState) || !showActiveState);
    const showIconOrLabel = !showAvatar && (showIcon || showLabel);

    const avatarAndLabel = (
      <>
        <span className={classNames(classes.avatar)}>{avatar}</span>
        {showLabel && <Label text={label} isForActiveState={!fullWidth && hasActiveState && isActive} />}
      </>
    );

    const iconAndLabel = (
      <>
        {showIcon && <Icon element={icon} elementClassName={iconClassName} />}
        {showLabel && <Label text={label} isForActiveState={!fullWidth && hasActiveState && isActive} />}
      </>
    );

    const activeStateContent = (
      <div className={classes.activeStateContent}>
        <Icon element={activeStateIcon} elementClassName={iconClassName} />
        <Label text={activeStateLabel} isForActiveState />
      </div>
    );

    return (
      <button type="button" {...rest} className={buttonClassName} ref={ref}>
        <div className={classes.buttonContent}>
          {showAvatar && avatarAndLabel}
          {showIconOrLabel && (fullWidth ? iconAndLabel : <span>{iconAndLabel}</span>)}
          {showActiveState && <Icon element={activeStateIcon} elementClassName={iconClassName} />}
          {showOpenButton && <IconAngleDown />}
          {showCollapseButton && <IconAngleUp />}
          {hasActiveState && activeStateContent}
        </div>
      </button>
    );
  },
);
