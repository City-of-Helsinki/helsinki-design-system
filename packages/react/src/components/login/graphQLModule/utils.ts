import { QueryOptions } from '@apollo/client';
import { merge } from 'lodash';

import { GraphQLModuleModuleProps } from '.';
import { ApiTokenClient } from '../apiTokensClient';

type PartialContext = Partial<QueryOptions['context']>;
type QueryHeaders = Record<string, unknown>;

function mergeQueryOptionsContexts(target: PartialContext, overrides: PartialContext): PartialContext {
  return merge(target, overrides);
}

export function mergeQueryOptionContexts(options: QueryOptions, overrideContext: PartialContext): QueryOptions {
  const context: PartialContext = options.context || {};
  // eslint-disable-next-line no-param-reassign
  options.context = mergeQueryOptionsContexts(context, overrideContext);
  return options;
}

export function mergeHeadersToQueryOptions(options: QueryOptions, headers: QueryHeaders): QueryOptions {
  const context: PartialContext = {
    headers,
  };
  return mergeQueryOptionContexts(options, context);
}

export function mergeAuthorizationHeaderToQueryOptions(options: QueryOptions, authHeader: string): QueryOptions {
  const headers: QueryHeaders = {
    authorization: authHeader,
  };
  return mergeHeadersToQueryOptions(options, headers);
}

export function mergeQueryOptions(
  target: Partial<GraphQLModuleModuleProps['queryOptions']>,
  overrides: Partial<GraphQLModuleModuleProps['queryOptions']>,
) {
  return merge(target, overrides);
}

/* No need to export to hds-react package */
export function mergeQueryOptionsToModuleProps<T, Q>(
  target: Partial<GraphQLModuleModuleProps<T, Q>>,
  queryOptions: Partial<QueryOptions<Q>>,
) {
  const { query, ...rest } = queryOptions;
  // eslint-disable-next-line no-param-reassign
  target.queryOptions = mergeQueryOptions(target.queryOptions || {}, rest);
  if (query) {
    // eslint-disable-next-line no-param-reassign
    target.query = query;
  }
  return target;
}

export function setBearerToQueryOptions(queryOptions: QueryOptions, token: string) {
  if (!token) {
    return queryOptions;
  }
  return mergeAuthorizationHeaderToQueryOptions(queryOptions, `Bearer ${token}`);
}

export function appendFetchOptions(queryOptions: QueryOptions, newOptions: Partial<Parameters<typeof fetch>[1]>) {
  if (!queryOptions.context) {
    // eslint-disable-next-line no-param-reassign
    queryOptions.context = {};
  }

  // eslint-disable-next-line no-param-reassign
  queryOptions.context.fetchOptions = {
    ...queryOptions.context.fetchOptions,
    ...newOptions,
  };
  return queryOptions;
}

/* No need to export to hds-react package */
export function mergeQueryOptionModifiers({
  options,
  queryHelper,
}: Pick<GraphQLModuleModuleProps, 'queryHelper' | 'options'>): Required<GraphQLModuleModuleProps>['queryHelper'] {
  const { apiTokenKey } = options || {};
  if (!apiTokenKey && !queryHelper) {
    return (opt: QueryOptions) => opt;
  }

  if (!apiTokenKey && queryHelper) {
    return queryHelper;
  }

  const tokenPicker = (apiTokenClient: ApiTokenClient, key: string) => {
    if (!apiTokenClient) {
      return undefined;
    }
    const tokens = apiTokenClient.getTokens();
    return tokens ? tokens[key] : undefined;
  };

  const apiTokenSetter: GraphQLModuleModuleProps['queryHelper'] = (opt, apiTokenClient) => {
    if (!apiTokenKey || !apiTokenClient) {
      return opt;
    }
    const token = tokenPicker(apiTokenClient, apiTokenKey);
    if (token) {
      return setBearerToQueryOptions(opt, token);
    }
    return opt;
  };

  if (!queryHelper) {
    return apiTokenSetter;
  }

  return (opt, apiTokenClient, beacon) => {
    return queryHelper(apiTokenSetter(opt, apiTokenClient, beacon), apiTokenClient, beacon);
  };
}
