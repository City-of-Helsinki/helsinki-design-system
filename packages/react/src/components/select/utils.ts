import React, { ReactElement, ReactNode } from 'react';

import {
  SelectData,
  Group,
  SelectProps,
  Option,
  OptionInProps,
  SelectMetaData,
  GroupInProps,
  FilterFunction,
} from './types';
import { getChildrenAsArray } from '../../utils/getChildren';
import { ChangeEvent } from '../dataProvider/DataContext';
import { eventTypes } from './events';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getOptionGroupIndex(groups: SelectData['groups'], option: OptionInProps): number {
  if (groups.length === 0) {
    return -1;
  }
  if (groups.length === 1) {
    return 0;
  }
  return groups.findIndex(({ options }) => {
    return (
      options.findIndex(({ value, isGroupLabel }) => value === option.value && isGroupLabel === option.isGroupLabel) >
      -1
    );
  });
}

export function iterateAndCopyGroup(groups: Group[], iterator: OptionIterator): Group[] {
  return groups.map((group, groupIndex) => {
    return {
      options: group.options.map((opt, optionIndex) => {
        return iterator(opt, group, optionIndex, groupIndex) || { ...opt };
      }),
    };
  });
}

export function getSelectedOptionsPerc(group: Group, pendingSelectionCount = 0): number {
  const optionCountWithoutGroupLabel = group.options.length - 1;
  if (!optionCountWithoutGroupLabel) {
    return 0;
  }
  return (
    (group.options.filter((option) => !option.isGroupLabel && option.selected).length + pendingSelectionCount) /
    optionCountWithoutGroupLabel
  );
}

export function updateOptionInGroup(
  groups: SelectData['groups'],
  updatedOption: Option,
  isMultiSelect: boolean,
): SelectData['groups'] {
  if (updatedOption.isGroupLabel) {
    throw new Error('Use updateGroupLabelAndOptions to update groupLabel and its related options');
  }
  const groupIndex = getOptionGroupIndex(groups, updatedOption);
  return groups.map((group, index) => {
    const selectedOptionPerc =
      index === groupIndex && isMultiSelect ? getSelectedOptionsPerc(group, updatedOption.selected ? 1 : -1) : 0;
    return {
      options: group.options.map((option) => {
        if (option.isGroupLabel) {
          if (index === groupIndex && isMultiSelect) {
            return {
              ...option,
              selected: selectedOptionPerc === 1,
            };
          }
          return { ...option };
        }
        if (index === groupIndex && option.value === updatedOption.value) {
          return {
            ...updatedOption,
            selected: !!updatedOption.selected,
          };
        }
        return {
          ...option,
          selected: !isMultiSelect ? false : option.selected,
        };
      }),
    };
  });
}

export function updateGroupLabelAndOptions(groups: SelectData['groups'], updatedOption: Option): SelectData['groups'] {
  if (!updatedOption.isGroupLabel) {
    throw new Error('Use updateOptionInGroup to update non-groupLabel options.');
  }
  if (!updatedOption.visible) {
    throw new Error('Cannot click an group label that is not visible (without a label)');
  }
  const targetGroupIndex = getOptionGroupIndex(groups, updatedOption);
  if (targetGroupIndex < 0) {
    return groups;
  }

  return iterateAndCopyGroup(groups, (option, group, optionIndex, groupIndex) => {
    if (groupIndex !== targetGroupIndex) {
      return { ...option };
    }
    return option.visible && !option.disabled
      ? {
          ...option,
          selected: updatedOption.selected,
        }
      : { ...option };
  });
}

export function clearAllSelectedOptions(groups: SelectData['groups']): SelectData['groups'] {
  return iterateAndCopyGroup(groups, (option) => {
    return {
      ...option,
      selected: false,
    };
  });
}

export function getAllOptions(groups: SelectData['groups'], filterOutGroupLabels = true): Option[] {
  const options: Option[] = [];
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (filterOutGroupLabels && option.isGroupLabel) {
        return;
      }
      options.push(option);
    });
  });
  return options;
}

export function getVisibleGroupLabels(groups: SelectData['groups']): Option[] {
  return groups.map((group) => group.options[0]).filter((option) => option && option.label && option.visible);
}

export function getSelectedOptions(groups: SelectData['groups']): Option[] {
  return getAllOptions(groups).filter((option) => !!option.selected);
}

export function validateOption(option: OptionInProps | string): Option {
  if (typeof option === 'string') {
    return { value: option, label: option, selected: false, isGroupLabel: false, visible: true, disabled: false };
  }

  const label = option.label || option.value || '';
  const value = option.value || label;

  return {
    label,
    value,
    selected: !!option.selected,
    isGroupLabel: false,
    visible: typeof option.visible === 'boolean' ? option.visible : true,
    disabled: typeof option.disabled === 'boolean' ? option.disabled : false,
  };
}

export function createGroupLabel(label: string) {
  return { ...validateOption(label), isGroupLabel: true, visible: !!label, disabled: false };
}

export function propsToGroups(props: Pick<SelectProps, 'groups' | 'options'>): SelectData['groups'] | undefined {
  if (!props.groups && !props.options) {
    return undefined;
  }
  if (props.groups) {
    return (props.groups as Group[]).map((group: Group) => {
      const hasLabelOptionAlready = !!group.options[0].isGroupLabel;
      const groupOptions = group.options.map(validateOption) as Option[];
      if (hasLabelOptionAlready) {
        return {
          options: [group.options[0], ...groupOptions.slice(1)],
        };
      }
      const labelOption: Option = createGroupLabel((group as GroupInProps).label);

      return {
        options: [labelOption, ...groupOptions],
      };
    });
  }

  return [
    {
      options: [createGroupLabel(''), ...(props.options || []).map(validateOption)],
    },
  ];
}

export function defaultFilter(option: Option, filterStr: string) {
  return option.label.toLowerCase().indexOf(filterStr.toLowerCase()) > -1;
}

export function filterOptions(groups: SelectData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function childrenToGroups(children: SelectProps['children']): SelectData['groups'] | undefined {
  if (!children || typeof children !== 'object') {
    return undefined;
  }
  const childArray = getChildrenAsArray(children) as ReactElement[];
  if (!childArray.length) {
    return [{ options: [] }];
  }
  const hasOptionGroups = childArray[0].type === 'optgroup';
  const optionElementToOption = (optionEl: ReactNode | ReactElement) => {
    const props = (
      optionEl && typeof optionEl === 'object' ? (optionEl as ReactElement).props : {}
    ) as ReactElement<HTMLOptionElement>['props'];
    const label = String(props.children);
    const value = props && String(props.value);
    const selected = !!(props && props.selected);
    const disabled = !!(props && props.disabled);
    return validateOption({ label, value, selected, disabled });
  };
  if (hasOptionGroups) {
    return childArray.map((child) => {
      const optionElements = child.props.children;
      const options = optionElements ? getChildrenAsArray(optionElements).map(optionElementToOption) : [];
      const label = createGroupLabel(String(child.props.label));
      options.unshift(label);
      return { options };
    });
  }
  return [{ options: [createGroupLabel(''), ...childArray.map(optionElementToOption)] }];
}

export function createSelectedOptionsList(currentSelections: Option[], groups: SelectData['groups']) {
  const selections = getSelectedOptions(groups);
  const selectionsAsValues = new Set(selections.map((opt) => opt.value));
  const stillSelected = currentSelections.filter((opt) => selectionsAsValues.has(opt.value));
  const selectedValues = new Set(stillSelected.map((opt) => opt.value));
  return [...stillSelected, ...selections.filter((opt) => !selectedValues.has(opt.value))];
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    arrowButton: `${containerId}-arrow-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function getGroupLabelOption(group: Group): Option | undefined {
  const firstOption = group.options[0];
  return firstOption && firstOption.isGroupLabel ? firstOption : undefined;
}
