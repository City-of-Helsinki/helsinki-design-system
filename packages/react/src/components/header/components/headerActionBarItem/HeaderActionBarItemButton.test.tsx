import React, { useRef } from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { HeaderActionBarItemButton, HeaderActionBarItemButtonProps } from './HeaderActionBarItemButton';
import { HeaderWrapper } from '../../../../utils/test-utils';
import { getCommonElementTestProps, getElementAttributesMisMatches } from '../../../../utils/testHelpers';
import { AllElementPropsWithRef } from '../../../../utils/elementTypings';

const clickHandler = jest.fn();

const defaultLabel = 'button label';
const fakeIconTestId = 'fake-icon-test-id';
const defaultIcon = <span data-testid={fakeIconTestId} />;
const activeStateLabel = 'close label';
const activeStateIconTestId = 'close-icon-test-id';
const activeStateIcon = <span data-testid={activeStateIconTestId} />;
const ariaControlsId = 'aria-controls-id';
const ariaLabel = 'aria label';

const defaultProps: HeaderActionBarItemButtonProps = {
  onClick: clickHandler,
  label: defaultLabel,
  icon: defaultIcon,
  'aria-expanded': false,
  'aria-label': ariaLabel,
  'aria-controls': ariaControlsId,
  labelOnRight: false,
  fixedRightPosition: false,
  isActive: false,
};

const activeStateProps: Pick<HeaderActionBarItemButtonProps, 'activeStateIcon' | 'activeStateLabel'> = {
  activeStateLabel,
  activeStateIcon,
};

const RenderTestScenario = (props?: Partial<HeaderActionBarItemButtonProps> & AllElementPropsWithRef<'button'>) => {
  const ref = useRef<HTMLButtonElement>(null);
  const combinedProps = { ...defaultProps, ...props };
  return <HeaderActionBarItemButton ref={ref} data-testid="button" {...combinedProps} />;
};

describe('HeaderActionBarItemButton spec', () => {
  it('renders the component correctly and adds all classNames from props', () => {
    const { asFragment } = render(
      <RenderTestScenario labelOnRight fixedRightPosition className="extra-class" {...activeStateProps} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(<RenderTestScenario />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it('native html props are passed to the element', async () => {
    const buttonProps = getCommonElementTestProps<'button'>('button');
    const { getByTestId } = render(<RenderTestScenario {...buttonProps} isActive labelOnRight fixedRightPosition />, {
      wrapper: HeaderWrapper,
    });
    const element = getByTestId(buttonProps['data-testid']);
    expect(getElementAttributesMisMatches(element, buttonProps)).toHaveLength(0);
  });

  it('Label and icon are set', async () => {
    const { getByText, getAllByTestId } = render(<RenderTestScenario />);
    expect(() => getByText(defaultLabel)).not.toThrow();
    expect(() => getAllByTestId(fakeIconTestId)).not.toThrow();
  });
  it('Aria attributes are set', async () => {
    const { getByTestId } = render(<RenderTestScenario aria-expanded />);
    const button = getByTestId('button') as HTMLElement;
    expect(button.getAttribute('aria-label')).toBe(ariaLabel);
    expect(button.getAttribute('aria-controls')).toBe(ariaControlsId);
    expect(button.getAttribute('aria-expanded')).toBe('true');
  });
  it('ActiveState content is rendered when "isActive" is true and "activeStateLabel" or "activeStateIcon" are passed', async () => {
    const { getAllByTestId } = render(<RenderTestScenario isActive {...activeStateProps} />);
    expect(() => getAllByTestId(activeStateIconTestId)).not.toThrow();
  });
  it('ActiveState content is also rendered when isActive is false and "activeStateLabel" or "activeStateIcon" are passed', async () => {
    const { getByText, getAllByTestId } = render(<RenderTestScenario {...activeStateProps} />);
    expect(() => getByText(defaultLabel)).not.toThrow();
    expect(() => getAllByTestId(fakeIconTestId)).not.toThrow();
  });
  it('ActiveState content is not rendered even if isActive is true, but neither "activeStateLabel" or "activeStateIcon" are passed', async () => {
    const { getByText, getAllByTestId } = render(<RenderTestScenario isActive />);
    expect(() => getByText(defaultLabel)).not.toThrow();
    expect(() => getAllByTestId(fakeIconTestId)).not.toThrow();
    expect(() => getByText(activeStateLabel)).toThrow();
    expect(() => getAllByTestId(activeStateIconTestId)).toThrow();
  });
});
