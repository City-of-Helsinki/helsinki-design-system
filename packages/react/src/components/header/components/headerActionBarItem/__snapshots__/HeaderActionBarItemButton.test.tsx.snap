// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderActionBarItemButton spec renders the component correctly and adds all classNames from props 1`] = `
<DocumentFragment>
  <button
    aria-controls="aria-controls-id"
    aria-expanded="false"
    aria-label="aria label"
    class="actionBarItemButton extra-class fixedRightPosition"
    data-testid="button"
    type="button"
  >
    <div
      class="buttonContent"
    >
      <span>
        <span
          class="actionBarItemButtonIcon labelOnRight"
        >
          <span
            aria-hidden="true"
            data-testid="fake-icon-test-id"
          />
        </span>
        <span
          class="actionBarItemButtonLabel labelOnRight"
        >
          button label
        </span>
      </span>
    </div>
    <div
      class="activeStateContent"
    >
      <span
        class="actionBarItemButtonIcon labelOnRight"
      >
        <span
          aria-hidden="true"
          data-testid="close-icon-test-id"
        />
      </span>
      <span
        aria-hidden="true"
        class="actionBarItemButtonLabel labelOnRight activeStateContentLabel"
      >
        close label
      </span>
    </div>
  </button>
</DocumentFragment>
`;
